File: cpp/tests/test_routing.cpp
#include "engine.hpp"
#include <cassert>
#include <iostream>
#include <cmath>

using namespace route_sim;

void test_routing_fixture(const std::string& fixture_path) {
    auto [graph, queries] = load_dataset(fixture_path);
    assert(graph.nodes.size() == 10);
    assert(graph.edges.size() == 26);
    assert(queries.size() == 6);

    for (const auto& q : queries) {
        auto res_d = solve_dijkstra(graph, q.source, q.destination, q.optimization);
        auto res_a = solve_astar(graph, q.source, q.destination, q.optimization);

        assert(res_d.has_value());
        assert(res_a.has_value());
        assert(std::abs(res_d->total_cost - res_a->total_cost) < 1e-4);
        assert(res_d->route == res_a->route);
    }
    std::cout << "[PASS] test_routing_fixture\n";
}

void test_edge_cases(const std::string& fixture_path) {
    auto [graph, queries] = load_dataset(fixture_path);

    // Unreachable
    auto res_unreachable = solve_dijkstra(graph, 0, 9999, "FASTEST");
    assert(!res_unreachable.has_value());

    // Same source and target
    auto res_same = solve_dijkstra(graph, 3, 3, "FASTEST");
    assert(res_same.has_value());
    assert(res_same->route.size() == 1 && res_same->route[0] == 3);
    assert(res_same->total_cost == 0.0);

    std::cout << "[PASS] test_edge_cases\n";
}

int main(int argc, char* argv[]) {
    std::string fixture_path = "../datasets/test_fixture.json";
    if (argc > 1) fixture_path = argv[1];

    std::cout << "--- Running C++ Routing Tests ---\n";
    test_routing_fixture(fixture_path);
    test_edge_cases(fixture_path);
    std::cout << "All C++ Routing Tests Passed Successfully!\n";
    return 0;
}

File: rust/tests/test_routing.rs
use route_optimization_rust::astar::solve_astar;
use route_optimization_rust::dijkstra::solve_dijkstra;
use route_optimization_rust::engine::load_dataset;

#[test]
fn test_fixture_routing() {
    let fixture_path = "../datasets/test_fixture.json";
    let (graph, queries) = load_dataset(fixture_path).expect("Failed to load test_fixture.json");

    assert_eq!(graph.nodes.len(), 10);
    assert_eq!(graph.edges.len(), 26);
    assert_eq!(queries.len(), 6);

    for q in &queries {
        let res_d = solve_dijkstra(&graph, q.source, q.destination, &q.optimization);
        let res_a = solve_astar(&graph, q.source, q.destination, &q.optimization);

        assert!(res_d.is_some());
        assert!(res_a.is_some());

        let d = res_d.unwrap();
        let a = res_a.unwrap();

        assert!((d.total_cost - a.total_cost).abs() < 1e-4);
        assert_eq!(d.route, a.route);
    }
}

#[test]
fn test_edge_cases() {
    let fixture_path = "../datasets/test_fixture.json";
    let (graph, _) = load_dataset(fixture_path).expect("Failed to load test_fixture.json");

    // Unreachable
    let res_unreachable = solve_dijkstra(&graph, 0, 9999, "FASTEST");
    assert!(res_unreachable.is_none());

    // Same source and target
    let res_same = solve_dijkstra(&graph, 3, 3, "FASTEST");
    assert!(res_same.is_some());
    let same = res_same.unwrap();
    assert_eq!(same.route, vec![3]);
    assert_eq!(same.total_cost, 0.0);
}
